onRenderBody.js

```diff
@@ -0,0 +1,48 @@
+import React from "react";
+import { DarkThemeKey, ThemeSetting } from "./src/theme/app/ThemeManager.js";
+import lighttheme, { darktheme } from "./src/theme/app/themeStyles";
+
+const themes = { light: lighttheme, dark: darktheme };
+
+const MagicScriptTag = (props) => {
+  const codeToRunOnClient = `
+      (function() {
+          const themeFromLocalStorage = localStorage.getItem('${DarkThemeKey}') || '${ThemeSetting.SYSTEM}';
+          const systemDarkModeSetting = () => window.matchMedia ? window.matchMedia('(prefers-color-scheme: dark)') : null;
+          const isDarkModeActive = () => {
+              return !!systemDarkModeSetting()?.matches;
+          };
+          let colorMode;
+          switch (themeFromLocalStorage) {
+              case '${ThemeSetting.SYSTEM}':
+                colorMode = isDarkModeActive() ? '${ThemeSetting.DARK}' : '{ThemeSetting.LIGHT}'
+                break
+              case '${ThemeSetting.DARK}':
+              case '${ThemeSetting.LIGHT}':
+                colorMode = themeFromLocalStorage
+                break
+              default:
+                  colorMode = '${ThemeSetting.LIGHT}'
+            }
+          const root = document.documentElement;
+          const iterate = (obj) => {
+              Object.keys(obj).forEach(key => {
+                 if (typeof obj[key] === 'object') {
+                      iterate(obj[key])
+                 } else {
+                   root.style.setProperty("--" + key, obj[key])
+                 }
+                 })
+          }
+          const parsedTheme = JSON.parse('${JSON.stringify(props.theme)}')
+          const theme = parsedTheme[colorMode]
+          iterate(theme)
+          root.style.setProperty('--initial-color-mode', colorMode);  
```

Comment:
Hi Nikhil,

Thanks for the comment and reviewing.

Please let me know if I have misunderstood your comment/request.

--------
I believe the onRenderBody code highlighted is necessary:
- to add CSS variable values to the root styles 
- so they can be referenced by a temporary CSS variable version of the default theme
- that is created in the StyledThemeProvider for the initial HTML created by Gatsby

My understanding is that this temporary CSS variable version of the theme is used while waiting for React to re-render, and is then replaced by the correct version of the theme.

If the "iterate" function is removed from onRenderBody, the flash returns, as the initialHTML will use the lighttheme by default:

![without_iterate](https://user-images.githubusercontent.com/90356410/227028732-e17a74dd-e1de-4dab-bac6-36fc62b95219.gif)

with the "iterate" function:

![with_iterate](https://user-images.githubusercontent.com/90356410/227034320-c52c1ed0-7371-40ea-8c20-ab5bbd8129cf.gif)

--------

in more detail:

The StyledThemeProvider component temporarily uses CSS variables to show the correct color mode in the initial HTML.
```
const theme = {
    ...(didLoad ? currentTheme : transformTheme(currentTheme)),
  };
```
Since didLoad is “false” in the initial HTML, "theme" will be the result of the "transformTheme" function. This function converts the values in the object into CSS-variables.
```
  Object.keys(theme).forEach((key) => {
    const value = theme[key];
    if (typeof value === "object" && !!value) {
      newTheme[key] = transformTheme(value);
    } else {
      newTheme[key] = `var(--${key})`;
    }
  });
```
example:
`background-color:  ${props => props.theme.body};`

will show in the initial HTML as:
`background-color:  var(--body)`

The background color will be whatever "--body" is set in the root style property. We set these root style properties in the onRenderBody function with the "iterate" function before the initial HTML is presented.

--------

Apologies if this explanation is not clear or I misunderstood your comment. Happy to discuss further.

Thanks,
Randy

Reply:
Hmm. interesting. I just noticed this behaviour.
Thanks for the explanation it helped :)